### DISCUSSION TITLE
Implementation of Recursive Largest First node coloring algorithm

### DISCUSSION THREAD

**Discussion by JonahBreslow** [networkx/networkx#6166] [Ideas]
I have created a python procedure that far outperforms all of the greedy coloring algorithms that are included in `networkx.algorithms.coloring` module. It is based on the [Recursive Largest First algorithm](https://en.wikipedia.org/wiki/Recursive_largest_first_algorithm), but it utilizes sampling to identify a maximal independent set instead of deterministically computing it over the graph. I have measured outperformance by comparing the results on the [discrete optimization coursera](https://www.coursera.org/learn/discrete-optimization) course. It is quite fast, as well. Would this be useful as a contribution to the coloring algorithms?

**Comment by dschult** [comment]
Yes, it would be a nice contribution to the coloring algorithms in NetworkX. 
I would prefer a version that could sample or deterministically find the maximal_independent_set (which we have a function for. That way, the user can choose whether to have the speed of the sampling method, or the exact nature of the solution for specific graph types like bipartitie.

If you can put together a PR that would be great!  If you just want to add it with sampling, that's fine. The deterministic way to get maximal_independent_sets can be added later.
Thanks!!

**Reply by JonahBreslow** [reply]
I do also have an implementation of the deterministic heuristic for finding the maximal independent set, but it is extremely slow at this point. I will attempt some optimizations and will try to add it in when I can improve performance. Thank you!

**Reply by dschult** [reply]
Is it using the NX function? I don't know how efficient ours is...   
I think though that you don't need to find the best maximal independent set. So long as it is a maximal independent set, it should work, right?

**Reply by JonahBreslow** [reply]
I believe the deterministic heuristic method to select the maximal independent set requires extra constraints that `nx.algorithms.mis` does not include. The algorithm for selecting the maximal independent set used in deterministic RLF is as follows:

1. `S` is an empty set
2. Select the node `n_max` where it is of the highest degree, and add it to `S`
3. Find the set of all nodes not adjacent to any node in `S`, call it `n_valid`
4. Find the set of node(s) in `n_valid` that have the maximal number of neighbors adjacent to `S`, call these `n_candidates`
5. if `n_candidates` has more than 1 node in it, select the node from `n_candidates` that has the minimum number of neighbors not in `S`
6. Repeat steps 3-5 until no more nodes can be added to `S`.

The NX function, `nx.algorithms.mis` is different:

1. `S` is an empty set 
2. We can select the node `n_max` where `n_max` is the node of the highest degree and add it to `S`
3. `nx.algorithms.mis` will find all nodes that are not neighbors with nodes in `S`, call it `n_candidates`
4. `nx.algorithms.mis` will randomly choose a node from `n_candidates` and will add it to `S`
5. Repeat steps 3-4 until a maximal independent set is constructed

While it is possible for `nx.algorithms.mis` to identify the same set as the deterministic heuristic method, there is no guarantee that it will. It will find _a_ maximal independent set, but not _the_ maximal independent set described in RLF. Particularly, steps 4 and 5 in the deterministic method are not present in the `nx.algorithms.mis` implementation, which is why it is so much faster consequently.

See ["A graph coloring algorithm for large scheduling problems"](https://nvlpubs.nist.gov/nistpubs/jres/84/jresv84n6p489_a1b.pdf)

**Reply by dschult** [reply]
It is true that there are multiple ways to generate maximal independent sets. They are "maximal", not "maximum". And it looks like even the implemented method is still a greedy algorithm that is unlikely to give the maximum.  But I'm assuming the authors of the paper checked what mis-algo tended to speed up the overall algorithm. They do suggest that the bottleneck for them is the mis-algo calculation.  There is a tradeoff between more time finding an indep set and getting fewer colors vs less time finding an indep set and needing to use more colors. 

Am I understanding that correctly?
